test/test_framework.py

```diff
@@ -1527,3 +1528,90 @@ def test_named_indicated_all(self, fake_stderr):
     def test_named_indicated_hook(self, fake_stderr):
         # The framework indicated the special value 'hook', nothing to do with any named call.
         self.check_trace_set('hook', 'mybreak', 0)
+
+
+@patch('sys.stderr', new_callable=io.StringIO)
+class DebugHookTests(unittest.TestCase):
+
+    def setUp(self):
+        tmpdir = Path(tempfile.mkdtemp())
+        self.addCleanup(shutil.rmtree, str(tmpdir))
+
+        self.framework = Framework(tmpdir / "framework.data", tmpdir, None, None)
+        self.addCleanup(self.framework.close)
+
+        class MyObserver(Object):
+            """Generic observer for the tests."""
+            def __init__(self, parent, key):
+                super().__init__(parent, key)
+                self.called = False
+
+            def callback_method(self, event):
+                self.called = True
+
+        self.charm_pub = charm.CharmEvents(self.framework, "1")
+        self.obs = MyObserver(self.framework, "1")
+
+    def test_basic_interruption_enabled(self, fake_stderr):
+        self.framework.observe(self.charm_pub.install, self.obs.callback_method)
+        with patch.dict(os.environ, {'JUJU_DEBUG_AT': 'hook'}):
+            with patch('pdb.runcall') as mock:
+                self.charm_pub.install.emit()
+
+        # Check that the pdb module was used correctly and that the callback method was NOT
+        # called (as we intercepted the normal pdb behaviour! this is to check that the
+        # framework didn't call the callback directly)
+        self.assertEqual(mock.call_count, 1)
+        expected_callback, expected_event = mock.call_args[0]
+        self.assertEqual(expected_callback, self.obs.callback_method)
+        self.assertIsInstance(expected_event, EventBase)
+        self.assertFalse(self.obs.called)
+
+    def test_internal_events_not_interrupted(self, fake_stderr):
+        class MyNotifier(Object):
+            """Generic notifier for the tests."""
+            bar = EventSource(EventBase)
+
+        pub = MyNotifier(self.framework, "1")
+        self.framework.observe(pub.bar, self.obs.callback_method)
+        with patch.dict(os.environ, {'JUJU_DEBUG_AT': 'hook'}):
```

Comment:
If you put it as an attribute of framework, I think you can test it cleaner than patching os.environ.

Reply:
Refactored the tests to use this good idea.